third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -34,30 +32,37 @@ use std::{
     sync::Arc,
 };
 
-pub struct AccountDataCache {
-    // The bool flag in the `data_map` indicates whether the resource contains
+struct AccountDataCache {
+    // The bool flag in the `resources` indicates whether the resource contains
     // an aggregator or snapshot.
-    data_map: BTreeMap<Type, (MoveTypeLayout, GlobalValue, bool)>,
-    module_map: BTreeMap<Identifier, (Bytes, bool)>,
+    resources: BTreeMap<Type, (MoveTypeLayout, GlobalValue, bool)>,
+    modules: BTreeMap<Identifier, ((Arc<CompiledModule>, Bytes), bool)>,
 }
 
 impl AccountDataCache {
     fn new() -> Self {
         Self {
-            data_map: BTreeMap::new(),
-            module_map: BTreeMap::new(),
+            resources: BTreeMap::new(),
+            modules: BTreeMap::new(),
         }
     }
 }
 
 fn load_module_impl(
-    remote: &dyn MoveResolver<PartialVMError>,
+    remote: &dyn MoveResolver<Arc<CompiledModule>, PartialVMError>,
     account_map: &BTreeMap<AccountAddress, AccountDataCache>,
     module_id: &ModuleId,
-) -> PartialVMResult<Bytes> {
+) -> PartialVMResult<(Arc<CompiledModule>, usize, [u8; 32])> {
     if let Some(account_cache) = account_map.get(module_id.address()) {
-        if let Some((blob, _is_republishing)) = account_cache.module_map.get(module_id.name()) {
-            return Ok(blob.clone());
+        if let Some(((m, b), _is_republishing)) = account_cache.modules.get(module_id.name()) {
+            // FIXME(George): Is it better to cache this information on publish?
+            //  In any case, once we move verification to deploy-time (and on storage
+            //  load in adapter), this can be removed.
+            let mut sha3_256 = Sha3_256::new();
+            sha3_256.update(b);
+            let hash_value: [u8; 32] = sha3_256.finalize().into();
+
+            return Ok((m.clone(), b.len(), hash_value));
```

Comment:
Hash of a module doesn't seem to me the information VM should care about?

Reply:
I need it for verification, because current verification cache lives in loader.rs ... otherwise yes, we do not need it.